--- a/extensions/libipt_MASQUERADE.c
+++ b/extensions/libipt_MASQUERADE.c
@@ -11,6 +11,7 @@
 enum {
 	O_TO_PORTS = 0,
 	O_RANDOM,
+	O_FULLCONE,
 	O_RANDOM_FULLY,
 };

@@ -23,13 +24,18 @@ static void MASQUERADE_help(void)
 " --random\n"
 "				Randomize source port.\n"
 " --random-fully\n"
-"				Fully randomize source port.\n");
+"				Fully randomize source port.\n"
+" --fullcone\n"
+"				SONiC-style fullcone NAT.\n");
 }

+#define NF_NAT_RANGE_FULLCONE (1 << 10)
+
 static const struct xt_option_entry MASQUERADE_opts[] = {
 	{.name = "to-ports", .id = O_TO_PORTS, .type = XTTYPE_STRING},
 	{.name = "random", .id = O_RANDOM, .type = XTTYPE_NONE},
 	{.name = "random-fully", .id = O_RANDOM_FULLY, .type = XTTYPE_NONE},
+	{.name = "fullcone", .id = O_FULLCONE, .type = XTTYPE_NONE},
 	XTOPT_TABLEEND,
 };

@@ -104,6 +109,9 @@ static void MASQUERADE_parse(struct xt_o
 	case O_RANDOM_FULLY:
 		mr->range[0].flags |=  NF_NAT_RANGE_PROTO_RANDOM_FULLY;
 		break;
+	case O_FULLCONE:
+		mr->range[0].flags |= NF_NAT_RANGE_FULLCONE;
+		break;
 	}
 }

@@ -126,6 +134,9 @@ MASQUERADE_print(const void *ip, const s

 	if (r->flags & NF_NAT_RANGE_PROTO_RANDOM_FULLY)
 		printf(" random-fully");
+
+	if (r->flags & NF_NAT_RANGE_FULLCONE)
+		printf(" fullcone");
 }

 static void
@@ -145,7 +156,10 @@ MASQUERADE_save(const void *ip, const st

 	if (r->flags & NF_NAT_RANGE_PROTO_RANDOM_FULLY)
 		printf(" --random-fully");
+
+	if (r->flags & NF_NAT_RANGE_FULLCONE)
+		printf(" --fullcone");
 }

 static int MASQUERADE_xlate(struct xt_xlate *xl,
